R/join_keys.R

```diff
@@ -0,0 +1,620 @@
+  original_value <- value
+  ds1 <- dataset_1
+
+  # In case an pair is removed, also remove the symmetric pair
+  removed_names <- setdiff(names(join_keys_obj[[dataset_1]]), names(value))
+  if (length(removed_names) > 0) {
+    for (.x in removed_names) {
+      value <- join_keys_obj[[.x]]
+      value[[ds1]] <- NULL
+      dataset_1 <- .x
+      join_keys_obj <- NextMethod("[[<-", join_keys_obj)
+    }
+
+    # Restore original values
+    dataset_1 <- ds1
+    value <- original_value
+  }
+
+  join_keys_obj <- NextMethod("[[<-", join_keys_obj)
+
+  # Iterate on all new values to create symmetrical pair
+  for (ds2 in names(original_value)) {
+    if (ds2 == ds1) next
+
+    value <- join_keys_obj[[ds2]] %||% list()
+    new_value <- original_value[[ds2]]
+
+    if (checkmate::test_character(new_value, min.len = 1, names = "unnamed")) {
+      new_value <- setNames(new_value, new_value)
+    } else if (
+      checkmate::test_character(new_value, min.len = 1)
+    ) {
+      # Invert key
+      new_value <- setNames(names(new_value), new_value)
+    }
+
+    # Change variables for NextMethod call
+    dataset_1 <- ds2
+    value[[ds1]] <- new_value
+    join_keys_obj <- NextMethod("[[<-", join_keys_obj)
+  }
+
+  join_keys_obj
+}
+
+#' @rdname merge_join_keys
+#' @keywords internal
+merge_join_keys <- function(join_keys_obj, new_join_keys) {
+  UseMethod("merge_join_keys", join_keys_obj)
+}
+
+#' @rdname merge_join_keys
+#' @keywords internal
+merge_join_keys.default <- function(join_keys_obj, new_join_keys) {
+  merge_join_keys(join_keys(join_keys_obj), new_join_keys)
+}
+
+#' Merging a list (or one) of `join_keys` objects into the current `join_keys` object
+#'
+#' @rdname merge_join_keys
+#'
+#' @param join_keys_obj (`join_keys`) object to merge the new_join_keys.
+#' @param new_join_keys  `list` of `join_keys` objects or single `join_keys` object
+#'
+#' @return a new `join_keys` object with the resulting merge.
+#'
+#' @keywords internal
+merge_join_keys.join_keys <- function(join_keys_obj, new_join_keys) {
+  if (
+    checkmate::test_class(new_join_keys, "join_key_set") ||
+      checkmate::test_class(new_join_keys, "join_keys")
+  ) {
+    new_join_keys <- list(new_join_keys)
+  }
+
+  lapply(new_join_keys, assert_join_keys_alike)
+
+  if (checkmate::test_list(new_join_keys, types = "join_key_set")) {
+    jk_temp <- new_join_keys()
+    join_keys(jk_temp) <- new_join_keys
+    new_join_keys <- list(jk_temp)
+  }
+
+  checkmate::assert_list(new_join_keys, types = c("join_keys"), min.len = 1)
+
+  for (el in new_join_keys) {
+    join_keys_obj <- utils::modifyList(join_keys_obj, el)
+  }
+
+  logger::log_trace("join_keys keys merged.")
+  return(join_keys_obj)
+}
+
+# S3 methods have to be exported, otherwise `.S3method` needs to be used
+.S3method("merge_join_keys", "teal_data", merge_join_keys.default)
+.S3method("merge_join_keys", "join_keys", merge_join_keys.join_keys)
```

Comment:
nice trick with `.S3method` 👍

Reply:
It's quite useful and only available from R >= 4 which is what this package supports :grin: